File: sdk/deepbook-v3/src/transactions/deepbookAdmin.ts
// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

import type { Transaction } from '@mysten/sui/transactions';

import type { CreatePoolAdminParams } from '../types/index.js';
import type { DeepBookConfig } from '../utils/config.js';
import { FLOAT_SCALAR } from '../utils/config.js';

/**
 * DeepBookAdminContract class for managing admin actions.
 */
export class DeepBookAdminContract {
	#config: DeepBookConfig;

	/**
	 * @param {DeepBookConfig} config Configuration for DeepBookAdminContract
	 */
	constructor(config: DeepBookConfig) {
		this.#config = config;
	}

	/**
	 * @returns The admin capability required for admin operations
	 * @throws Error if the admin capability is not set
	 */
	#adminCap() {
		const adminCap = this.#config.adminCap;
		if (!adminCap) {
			throw new Error('ADMIN_CAP environment variable not set');
		}
		return adminCap;
	}

	/**
	 * @description Create a new pool as admin
	 * @param {CreatePoolAdminParams} params Parameters for creating pool as admin
	 * @returns A function that takes a Transaction object
	 */
	createPoolAdmin = (params: CreatePoolAdminParams) => (tx: Transaction) => {
		tx.setSenderIfNotSet(this.#config.address);
		const { baseCoinKey, quoteCoinKey, tickSize, lotSize, minSize, whitelisted, stablePool } =
			params;
		const baseCoin = this.#config.getCoin(baseCoinKey);
		const quoteCoin = this.#config.getCoin(quoteCoinKey);

		const baseScalar = baseCoin.scalar;
		const quoteScalar = quoteCoin.scalar;

		const adjustedTickSize = (tickSize * FLOAT_SCALAR * quoteScalar) / baseScalar;
		const adjustedLotSize = lotSize * baseScalar;
		const adjustedMinSize = minSize * baseScalar;

		tx.moveCall({
			target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::create_pool_admin`,
			arguments: [
				tx.object(this.#config.REGISTRY_ID), // registry_id
				tx.pure.u64(adjustedTickSize), // adjusted tick_size
				tx.pure.u64(adjustedLotSize), // adjusted lot_size
				tx.pure.u64(adjustedMinSize), // adjusted min_size
				tx.pure.bool(whitelisted),
				tx.pure.bool(stablePool),
				tx.object(this.#adminCap()),
			],
			typeArguments: [baseCoin.type, quoteCoin.type],
		});
	};

	/**
	 * @description Unregister a pool as admin
	 * @param {string} poolKey The key of the pool to be unregistered by admin
	 * @returns A function that takes a Transaction object
	 */
	unregisterPoolAdmin = (poolKey: string) => (tx: Transaction) => {
		const pool = this.#config.getPool(poolKey);
		const baseCoin = this.#config.getCoin(pool.baseCoin);
		const quoteCoin = this.#config.getCoin(pool.quoteCoin);
		tx.moveCall({
			target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::unregister_pool_admin`,
			arguments: [
				tx.object(pool.address),
				tx.object(this.#config.REGISTRY_ID),
				tx.object(this.#adminCap()),
			],
			typeArguments: [baseCoin.type, quoteCoin.type],
		});
	};

	/**
	 * @description Update the allowed versions for a pool
	 * @param {string} poolKey The key of the pool to be updated
	 * @returns A function that takes a Transaction object
	 */
	updateAllowedVersions = (poolKey: string) => (tx: Transaction) => {
		const pool = this.#config.getPool(poolKey);
		const baseCoin = this.#config.getCoin(pool.baseCoin);
		const quoteCoin = this.#config.getCoin(pool.quoteCoin);
		tx.moveCall({
			target: `${this.#config.DEEPBOOK_PACKAGE_ID}::pool::update_allowed_versions`,
			arguments: [
				tx.object(pool.address),
				tx.object(this.#config.REGISTRY_ID),
				tx.object(this.#adminCap()),
			],
			typeArguments: [baseCoin.type, quoteCoin.type],
		});
	};

	/**
	 * @description Enable a specific version
	 * @param {number} version The version to be enabled
	 * @returns A function that takes a Transaction object
	 */
	enableVersion = (version: number) => (tx: Transaction) => {
		tx.moveCall({
			target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::enable_version`,
			arguments: [
				tx.object(this.#config.REGISTRY_ID),
				tx.pure.u64(version),
				tx.object(this.#adminCap()),
			],
		});
	};

	/**
	 * @description Disable a specific version
	 * @param {number} version The version to be disabled
	 * @returns A function that takes a Transaction object
	 */
	disableVersion = (version: number) => (tx: Transaction) => {
		tx.moveCall({
			target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::disable_version`,
			arguments: [
				tx.object(this.#config.REGISTRY_ID),
				tx.pure.u64(version),
				tx.object(this.#adminCap()),
			],
		});
	};

	/**
	 * @description Sets the treasury address where pool creation fees will be sent
	 * @param {string} treasuryAddress The treasury address
	 * @returns A function that takes a Transaction object
	 */
	setTreasuryAddress = (treasuryAddress: string) => (tx: Transaction) => {
		tx.moveCall({
			target: `${this.#config.DEEPBOOK_PACKAGE_ID}::registry::set_treasury_address`,
			arguments: [
				tx.object(this.#config.REGISTRY_ID),
				tx.pure.address(treasuryAddress),
				tx.object(this.#adminCap()),
			],
		});
	};
}

File: crates/sui-deepbookv3/src/transactions/deepbook_admin.rs
// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::str::FromStr;

use anyhow::Result;
use sui_sdk::{
    types::{
        base_types::{ObjectID, SuiAddress},
        programmable_transaction_builder::ProgrammableTransactionBuilder,
        Identifier, TypeTag,
    },
    SuiClient,
};

use crate::utils::config::{DeepBookConfig, FLOAT_SCALAR};

use crate::DataReader;

/// Parameters for creating a pool as admin
pub struct CreatePoolAdminParams {
    pub base_coin_key: String,
    pub quote_coin_key: String,
    pub tick_size: f64,
    pub lot_size: f64,
    pub min_size: f64,
    pub whitelisted: bool,
    pub stable_pool: bool,
}

/// DeepBookAdminContract struct for managing admin actions
pub struct DeepBookAdminContract {
    client: SuiClient,
    config: DeepBookConfig,
}

impl DeepBookAdminContract {
    /// Creates a new DeepBookAdminContract instance
    ///
    /// @param client - SuiClient instance
    /// @param config - Configuration for DeepBookAdminContract
    pub fn new(client: SuiClient, config: DeepBookConfig) -> Self {
        Self { client, config }
    }

    /// Gets the admin capability required for admin operations
    ///
    /// @returns The admin capability
    /// @throws Error if the admin capability is not set
    fn admin_cap(&self) -> anyhow::Result<String> {
        self.config
            .admin_cap()
            .ok_or_else(|| anyhow::anyhow!("ADMIN_CAP environment variable not set"))
    }

    /// Create a new pool as admin
    ///
    /// @param ptb - ProgrammableTransactionBuilder instance
    /// @param params - Parameters for creating pool as admin
    pub async fn create_pool_admin(
        &self,
        ptb: &mut ProgrammableTransactionBuilder,
        params: CreatePoolAdminParams,
    ) -> Result<()> {
        let base_coin = self.config.get_coin(&params.base_coin_key)?;
        let quote_coin = self.config.get_coin(&params.quote_coin_key)?;

        let base_scalar = base_coin.scalar;
        let quote_scalar = quote_coin.scalar;

        let adjusted_tick_size =
            (params.tick_size * FLOAT_SCALAR as f64 * quote_scalar as f64) / base_scalar as f64;
        let adjusted_lot_size = params.lot_size * base_scalar as f64;
        let adjusted_min_size = params.min_size * base_scalar as f64;

        let base_coin_tag = TypeTag::from_str(&base_coin.type_name)?;
        let quote_coin_tag = TypeTag::from_str(&quote_coin.type_name)?;

        let registry_id = ObjectID::from_hex_literal(self.config.registry_id())?;
        let admin_cap = ObjectID::from_hex_literal(&self.admin_cap()?)?;
        let arguments = vec![
            ptb.obj(self.client.share_object(registry_id).await?)?,
            ptb.pure(adjusted_tick_size as u64)?,
            ptb.pure(adjusted_lot_size as u64)?,
            ptb.pure(adjusted_min_size as u64)?,
            ptb.pure(params.whitelisted)?,
            ptb.pure(params.stable_pool)?,
            ptb.obj(self.client.share_object(admin_cap).await?)?,
        ];

        ptb.programmable_move_call(
            ObjectID::from_hex_literal(self.config.deepbook_package_id())?,
            Identifier::new("pool")?,
            Identifier::new("create_pool_admin")?,
            vec![base_coin_tag, quote_coin_tag],
            arguments,
        );

        Ok(())
    }

    /// Unregister a pool as admin
    ///
    /// @param ptb - ProgrammableTransactionBuilder instance
    /// @param pool_key - The key of the pool to be unregistered by admin
    pub async fn unregister_pool_admin(
        &self,
        ptb: &mut ProgrammableTransactionBuilder,
        pool_key: &str,
    ) -> Result<()> {
        let pool = self.config.get_pool(pool_key)?;
        let base_coin = self.config.get_coin(&pool.base_coin)?;
        let quote_coin = self.config.get_coin(&pool.quote_coin)?;

        let base_coin_tag = TypeTag::from_str(&base_coin.type_name)?;
        let quote_coin_tag = TypeTag::from_str(&quote_coin.type_name)?;

        let pool_id = ObjectID::from_hex_literal(&pool.address)?;
        let registry_id = ObjectID::from_hex_literal(self.config.registry_id())?;
        let admin_cap = ObjectID::from_hex_literal(&self.admin_cap()?)?;

        let arguments = vec![
            ptb.obj(self.client.share_object(pool_id).await?)?,
            ptb.obj(self.client.share_object(registry_id).await?)?,
            ptb.obj(self.client.share_object(admin_cap).await?)?,
        ];

        ptb.programmable_move_call(
            ObjectID::from_hex_literal(self.config.deepbook_package_id())?,
            Identifier::new("pool")?,
            Identifier::new("unregister_pool_admin")?,
            vec![base_coin_tag, quote_coin_tag],
            arguments,
        );

        Ok(())
    }

    /// Update the allowed versions for a pool
    ///
    /// @param ptb - ProgrammableTransactionBuilder instance
    /// @param pool_key - The key of the pool to be updated
    pub async fn update_allowed_versions(
        &self,
        ptb: &mut ProgrammableTransactionBuilder,
        pool_key: &str,
    ) -> Result<()> {
        let pool = self.config.get_pool(pool_key)?;
        let base_coin = self.config.get_coin(&pool.base_coin)?;
        let quote_coin = self.config.get_coin(&pool.quote_coin)?;

        let base_coin_tag = TypeTag::from_str(&base_coin.type_name)?;
        let quote_coin_tag = TypeTag::from_str(&quote_coin.type_name)?;

        let pool_id = ObjectID::from_hex_literal(&pool.address)?;
        let registry_id = ObjectID::from_hex_literal(self.config.registry_id())?;
        let admin_cap = ObjectID::from_hex_literal(&self.admin_cap()?)?;

        let arguments = vec![
            ptb.obj(self.client.share_object(pool_id).await?)?,
            ptb.obj(self.client.share_object(registry_id).await?)?,
            ptb.obj(self.client.share_object(admin_cap).await?)?,
        ];

        ptb.programmable_move_call(
            ObjectID::from_hex_literal(self.config.deepbook_package_id())?,
            Identifier::new("pool")?,
            Identifier::new("update_allowed_versions")?,
            vec![base_coin_tag, quote_coin_tag],
            arguments,
        );

        Ok(())
    }

    /// Enable a specific version
    ///
    /// @param ptb - ProgrammableTransactionBuilder instance
    /// @param version - The version to be enabled
    pub async fn enable_version(
        &self,
        ptb: &mut ProgrammableTransactionBuilder,
        version: u64,
    ) -> Result<()> {
        let registry_id = ObjectID::from_hex_literal(self.config.registry_id())?;
        let admin_cap = ObjectID::from_hex_literal(&self.admin_cap()?)?;

        let arguments = vec![
            ptb.obj(self.client.share_object(registry_id).await?)?,
            ptb.pure(version)?,
            ptb.obj(self.client.share_object(admin_cap).await?)?,
        ];

        ptb.programmable_move_call(
            ObjectID::from_hex_literal(self.config.deepbook_package_id())?,
            Identifier::new("registry")?,
            Identifier::new("enable_version")?,
            vec![],
            arguments,
        );

        Ok(())
    }

    /// Disable a specific version
    ///
    /// @param ptb - ProgrammableTransactionBuilder instance
    /// @param version - The version to be disabled
    pub async fn disable_version(
        &self,
        ptb: &mut ProgrammableTransactionBuilder,
        version: u64,
    ) -> Result<()> {
        let registry_id = ObjectID::from_hex_literal(self.config.registry_id())?;
        let admin_cap = ObjectID::from_hex_literal(&self.admin_cap()?)?;

        let arguments = vec![
            ptb.obj(self.client.share_object(registry_id).await?)?,
            ptb.pure(version)?,
            ptb.obj(self.client.share_object(admin_cap).await?)?,
        ];

        ptb.programmable_move_call(
            ObjectID::from_hex_literal(self.config.deepbook_package_id())?,
            Identifier::new("registry")?,
            Identifier::new("disable_version")?,
            vec![],
            arguments,
        );

        Ok(())
    }

    /// Sets the treasury address where pool creation fees will be sent
    ///
    /// @param ptb - ProgrammableTransactionBuilder instance
    /// @param treasury_address - The treasury address
    pub async fn set_treasury_address(
        &self,
        ptb: &mut ProgrammableTransactionBuilder,
        treasury_address: &str,
    ) -> Result<()> {
        let registry_id = ObjectID::from_hex_literal(self.config.registry_id())?;
        let admin_cap = ObjectID::from_hex_literal(&self.admin_cap()?)?;

        let arguments = vec![
            ptb.obj(self.client.share_object(registry_id).await?)?,
            ptb.pure(SuiAddress::from_str(treasury_address)?)?,
            ptb.obj(self.client.share_object(admin_cap).await?)?,
        ];

        ptb.programmable_move_call(
            ObjectID::from_hex_literal(self.config.deepbook_package_id())?,
            Identifier::new("registry")?,
            Identifier::new("set_treasury_address")?,
            vec![],
            arguments,
        );

        Ok(())
    }
}
